### ISSUE TITLE
Breaking change on getters being applied for populate()

### ISSUE THREAD

**Issue by daltones** [Automattic/mongoose#6844]
@lineus, @trungtin and @vkarpov15,

I'm opening this issue as an attempt to clarify things about #6618 and #6702 [and consequent #6726 and #6729]. Right now my app is stuck in  mongoose v5.2.2 and only today I had the time to try to understand what is breaking my code.

#### Summarizing
I came to the conclusion that #6618 **was not a `confirmed-bug`**. It's just a **different use case that does not fit in "getters" feature**. This misconception led us to #6702, which basically forced getters/populate to fit in that use case, at the cost of **breaking the expected reference system of `populate()`**.

#### Arguing
I've been using mongoose for a while, and AFAIK I would describe "getters" and "populate" as:
- `populate`: it's mongoose's way to simulate SQL's `JOIN` (a document has a property that _references_ another document).
- `getter`: it's a filter to be applied on the raw document value (fetched from the database) when someone accesses a model property.

Now, considering `populate`/`JOIN`: we expect that a local value matches _exactly_ a foreign value, right? At least _we currently do not have the feature of specifying a custom matching rule for local/foreign values_.

I'll show a use case of getters + populate that used to work before and it's now broken:

```js
const stateSchema = new Schema({
  _id: String,
  name: String,
});

const citySchema = new Schema({
  _id: String,
  name: String,
  state: {
    type: String,
    get: value => `StateId:${value}`,
    ref: 'State',
  },
});

/*
Let's assume we have these documents persisted in our database:

Collection "states":
{"_id": "s1", "name": "California"}

Collection "cities":
{"_id": "c1", "name": "Los Angeles", "state": "s1"}

Seem's a fine relationship, right?
*/


const losAngeles = await City.findById('c1');
console.log(losAngeles.state);
// prints "StateId:s1"
// good, just filtering the fetched value from the document due to our defined getter for "state".


const losAngelesPopulatedState = await City.findById('c1').populate('state', 'name');
console.log(losAngelesPopulatedState.state._id);
// prints "s1"
// we expect that this population did work
```

The last `populate`/`JOIN` does not work after mongoose v5.2.2 :disappointed:

That's because we wrongly applied the getter on the value that we expect to match exactly with the foreign value.

Looking forward on solving this issue. Thanks!

**Comment by lineus** [comment]
Thanks for reporting this @daltones! My apologies for the backwards breaking change.

# Replicated with:

### 6844.js
```js
#!/usr/bin/env node
'use strict';

const assert = require('assert');
const mongoose = require('mongoose');
mongoose.connect('mongodb://localhost:27017/test', { useNewUrlParser: true });
const conn = mongoose.connection;
const Schema = mongoose.Schema;

const stateSchema = new Schema({
  _id: String,
  name: String
});

const citySchema = new Schema({
  _id: String,
  name: String,
  state: {
    type: String,
    get: value => `StateId:${value}`,
    ref: 'State'
  }
});

const State = mongoose.model('State', stateSchema);
const City = mongoose.model('City', citySchema);

const cal = { _id: 's1', name: 'California' };
const la = { _id: 'c1', name: 'Los Angeles', state: 's1' };

async function run() {
  await conn.dropDatabase();
  await State.create(cal);
  let saved = await City.create(la);
  assert.strictEqual(saved.state, 'StateId:s1', 'accessing property fails');
  
  let doc = await City.findById('c1').populate('state');
  assert.strictEqual(doc.state.name, 'California', 'population fails');
  
  return conn.close();
}

run().catch((e) => { 
  console.error(e);
  return mongoose.disconnect(); 
});
```
### Output:
```
issues: ./6844.js
{ AssertionError [ERR_ASSERTION]: population fails
    at run (/Users/lineus/dev/Help/mongoose5/issues/6844.js:38:10)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
  generatedMessage: false,
  name: 'AssertionError [ERR_ASSERTION]',
  code: 'ERR_ASSERTION',
  actual: undefined,
  expected: 'California',
  operator: '===' }
issues:
```

# One potential userland solution:
@daltones With the current version of mongoose you can change your getter to the following:
```
    get: function(value) { return this.$init ? value:`StateId:${value}`; },
```
and the output of the above repro code becomes:
```
issues: ./6844.js
issues:
```

# Thoughts

Personally, I think running the getter on the path during population adds enough value to keep it around, but I also recognize that the behavior changed in a patch release.

**Is an option the way to go here?**

Something like

```
const citySchema = new Schema({
  _id: String,
  name: String,
  state: {
    type: String,
    get: value => `StateId:${value}`,
    runGetterWhenPopulating: true, // defaults to undefined
    ref: 'State'
  }
});
```

@trungtin, this would mean that you would need to set this option for your use case. 

@vkarpov15, @trungtin, @daltones what do you guys think of this solution?

**Comment by vkarpov15** [comment]
You're right, this was not intended to be a backwards breaking change but we didn't think of this use case. @daltones ' point is solid. I like @lineus ' idea of adding an option, How about a `getters` option that you can set either in `schema.virtual(path, options)` or in `query.populate({ path, options })` to opt in to calling getters?

@lineus your proposed userland solution works for this case, but not necessarily all cases. I don't think it'll work if you use [`doc.populate()`](https://mongoosejs.com/docs/api.html#document_Document-populate) as opposed to [`query.populate()`](https://mongoosejs.com/docs/api.html#query_Query-populate) for instance, because in `doc.populate()` the doc is already hydrated when the getter gets called.

**Comment by lineus** [comment]
@vkarpov15 that makes sense 👍

**Comment by vkarpov15** [comment]
Fix will be in 5.2.10, getters won't be applied unless you explicitly set `getters: true` in `schema.virtual()` or your `populate()` call:

```javascript
schema.virtual('foo', {
  ref: 'Bar',
  localField: 'baz',
  foreignField: '_id',
  getters: true // Opt in to calling getters on 'baz'
});
```

```javascript
